pyls/rpc_manager.py

```diff
@@ -0,0 +1,190 @@
+        try:
+            maybe_handler = self._message_handler(request.method, request.params if request.params is not None else {})
+        except MissingMethodException as e:
+            log.debug(e)
+            # Do not need to notify client of failure with notifications
+            output = JSONRPC20Response(_id=request._id, error=JSONRPCMethodNotFound()._data)
+        except JSONRPCDispatchException as e:
+            output = _make_response(request, error=e.error._data)
+        except Exception as e:  # pylint: disable=broad-except
+            log.error('endpoint exception %s %s %s', e.__class__.__name__, e.args, str(e))
+            output = _make_response(request, error=JSONRPCServerError()._data)
+        else:
+            if request._id in self._received_requests:
+                log.error('Received request %s with duplicate id', request.data)
+            elif callable(maybe_handler):
+                log.debug('Async request %s', request._id)
+                self._handle_async_request(request, maybe_handler)
+            else:
+                output = _make_response(request, result=maybe_handler)
+        finally:
+            if not request.is_notification and output is not None:
+                log.debug('Sync request %s', request._id)
+                self._message_manager.write_message(output)
+
+    def _handle_async_request(self, request, handler):
+        future = self._executor_service.submit(handler)
+
+        if request.is_notification:
+            return
+
+        def did_finish_callback(completed_future):
+            del self._received_requests[request._id]
+            if completed_future.cancelled():
+                log.debug('Cleared cancelled request %d', request._id)
+            else:
+                try:
+                    result = completed_future.result()
+                except JSONRPCDispatchException as e:
+                    output = _make_response(request, error=e.error._data)
+                except Exception as e:  # pylint: disable=broad-except
+                    # TODO(forozco): add more descriptive error
+                    log.error('endpoint exception %s %s %s', e.__class__.__name__, e.args, str(e))
+                    output = _make_response(request, error=JSONRPCServerError()._data)
+                else:
+                    output = _make_response(request, result=result)
+                finally:
+                    self._message_manager.write_message(output)
+
+        self._received_requests[request._id] = future
+        future.add_done_callback(did_finish_callback)
+
+    def _handle_response(self, response):
+        """Handle the response to requests sent from the server to the client.
+
+        Args:
+            response: (JSONRPC20Response): Received response
+
+        """
+        try:
+            request = self._sent_requests[response._id]
+        except KeyError:
+            log.error('Received unexpected response %s', response.data)
```

Comment:
I usually prefer an explicit check for these sorts of things `if response._id not in self._sent_requests`. Particularly in this case where the try/else block makes up the bulk of the function.